File: src/main/java/com/vaadin/demo/component/datetimepicker/DateTimePickerCustomValidation.java
package com.vaadin.demo.component.datetimepicker;

import com.vaadin.flow.component.datetimepicker.DateTimePicker;
import com.vaadin.flow.component.html.Div;
import com.vaadin.flow.data.binder.Binder;
import com.vaadin.flow.router.Route;

import java.time.Duration;
import java.time.LocalTime;

import com.vaadin.demo.DemoExporter; // hidden-source-line
import com.vaadin.demo.domain.Appointment;

@Route("date-time-picker-custom-validation")
public class DateTimePickerCustomValidation extends Div {

    public DateTimePickerCustomValidation() {
        // tag::snippet[]
        DateTimePicker dateTimePicker = new DateTimePicker();
        dateTimePicker.setLabel("Appointment date and time");
        dateTimePicker
                .setHelperText("Open Mondays-Fridays, 8:00-12:00, 13:00-16:00");
        dateTimePicker.setStep(Duration.ofMinutes(30));
        add(dateTimePicker);

        String errorMessage = "The selected day of week or time is not available";
        Binder<Appointment> binder = new Binder<>(Appointment.class);
        binder.forField(dateTimePicker).withValidator(startDateTime -> {
            boolean validWeekDay = startDateTime.getDayOfWeek().getValue() >= 1
                    && startDateTime.getDayOfWeek().getValue() <= 5;
            return validWeekDay;
        }, errorMessage).withValidator(startDateTime -> {
            LocalTime startTime = LocalTime.of(startDateTime.getHour(),
                    startDateTime.getMinute());
            boolean validTime = !(LocalTime.of(8, 0).isAfter(startTime)
                    || (LocalTime.of(12, 0).isBefore(startTime)
                            && LocalTime.of(13, 0).isAfter(startTime))
                    || LocalTime.of(16, 0).isBefore(startTime));
            return validTime;
        }, errorMessage).bind(Appointment::getStartDateTime,
                Appointment::setStartDateTime);
        // end::snippet[]
    }

    public static class Exporter extends // hidden-source-line
            DemoExporter<DateTimePickerCustomValidation> { // hidden-source-line
    } // hidden-source-line
}

File: frontend/demo/component/datetimepicker/date-time-picker-custom-validation.ts
import 'Frontend/demo/init'; // hidden-source-line

import { html, LitElement } from 'lit';
import { customElement } from 'lit/decorators.js';
import '@vaadin/vaadin-date-time-picker/vaadin-date-time-picker';
import { Binder, field } from '@vaadin/form';
import AppointmentModel from 'Frontend/generated/com/vaadin/demo/domain/AppointmentModel';
import { applyTheme } from 'Frontend/generated/theme';

@customElement('date-time-picker-custom-validation')
export class Example extends LitElement {
  protected createRenderRoot() {
    const root = super.createRenderRoot();
    // Apply custom theme (only supported if your app uses one)
    applyTheme(root);
    return root;
  }

  // tag::snippet[]
  private binder = new Binder(this, AppointmentModel);

  firstUpdated() {
    this.binder.for(this.binder.model.startDateTime).addValidator({
      message: 'The selected day of week is not available',
      validate: (startDateTime: string) => {
        const date = new Date(startDateTime);
        const validWeekDay = date.getDay() >= 1 && date.getDay() <= 5;
        return validWeekDay;
      },
    });
    this.binder.for(this.binder.model.startDateTime).addValidator({
      message: 'The selected time is not available',
      validate: (startDateTime: string) => {
        const time = startDateTime.split('T')[1];
        const validTime =
          (time >= '08:00' && time <= '12:00') || (time >= '13:00' && time <= '16:00');
        return validTime;
      },
    });
  }

  render() {
    return html`
      <vaadin-date-time-picker
        label="Appointment date and time"
        helper-text="Open Mondays-Fridays, 8:00-12:00, 13:00-16:00"
        .step="${60 * 30}"
        ...="${field(this.binder.model.startDateTime)}"
      ></vaadin-date-time-picker>
    `;
  }
  // end::snippet[]
}
